### PR TITLE
Use 128bit instead of 64bit amounts

### LINKED ISSUE #1017: Handle new circuit that supports amount > 2**63
See https://github.com/EspressoSystems/cap/issues/37

- [ ] Point this repository to latest version of Jellyfish
- [ ] Regenerate the smart contract code corresponding to the new plonk verifying keys. See See https://github.com/EspressoSystems/cap/issues/37#issuecomment-1134733906 fo
- [ ] Test that ensures amount > 2**63 can be handled

### PR DESCRIPTION
**Pull request by sveitser**
Close #1017 .

Still requires
- https://github.com/EspressoSystems/seahorse/pull/169

to be merged, followed by changing back the seahorse branch in the `Cargo.toml` files.

### CODE REVIEW

**Comment by sveitser** [comment]
After rebasing on `main` with the faucet improvement I get this compilation error

```
$  cargo test --no-run                                                                                                               70.642s  feat/t1017-circuit-128-bit  ❄  16:40:15 
   Compiling cap-rust-sandbox v0.2.0 (/home/lulu/r/EspressoSystems/cape/contracts/rust)
   Compiling eqs v0.0.2 (/home/lulu/r/EspressoSystems/cape/eqs)
   Compiling relayer v0.2.0 (/home/lulu/r/EspressoSystems/cape/relayer)
   Compiling cape_wallet v0.2.0 (/home/lulu/r/EspressoSystems/cape/wallet)
error: lifetime may not live long enough
   --> wallet/src/testing.rs:103:3
    |
93  |   pub async fn create_test_network<'a>(
    |                                    -- lifetime `'a` defined here
94  |       rng: &mut ChaChaRng,
    |            - let's call the lifetime of this reference `'1`
...
103 |   ) {
    |  ___^
104 | |     init_web_server(TransientFileStore::default())
105 | |         .await
106 | |         .expect("Failed to run server.");
...   |
136 | |     )
137 | | }
    | |_^ function was supposed to return data with lifetime `'a` but it is returning data with lifetime `'1`
    |
    = note: requirement occurs because of the type async_std::sync::Mutex<MockLedger<'_, cap_rust_sandbox::ledger::CapeLedger, mocks::MockCapeNetwork, AtomicWalletStorage<'_, cap_rust_sandbox::ledger::CapeLedger, ()>>>, which makes the generic argument MockLedger<'_, cap_rust_sandbox::ledger::CapeLedger, mocks::MockCapeNetwork, AtomicWalletStorage<'_, cap_rust_sandbox::ledger::CapeLedger, ()>> invariant
    = note: the struct async_std::sync::Mutex<T> is invariant over the parameter T
    = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance

error: lifetime may not live long enough
   --> wallet/src/testing.rs:103:3
    |
93  |   pub async fn create_test_network<'a>(
    |                                    -- lifetime `'a` defined here
94  |       rng: &mut ChaChaRng,
    |            - let's call the lifetime of this reference `'1`
...
103 |   ) {
    |  ___^
104 | |     init_web_server(TransientFileStore::default())
105 | |         .await
106 | |         .expect("Failed to run server.");
...   |
136 | |     )
137 | | }
    | |_^ function was supposed to return data with lifetime `'1` but it is returning data with lifetime `'a`
    |
    = note: requirement occurs because of the type async_std::sync::Mutex<MockLedger<'_, cap_rust_sandbox::ledger::CapeLedger, mocks::MockCapeNetwork, AtomicWalletStorage<'_, cap_rust_sandbox::ledger::CapeLedger, ()>>>, which makes the generic argument MockLedger<'_, cap_rust_sandbox::ledger::CapeLedger, mocks::MockCapeNetwork, AtomicWalletStorage<'_, cap_rust_sandbox::ledger::CapeLedger, ()>> invariant
    = note: the struct async_std::sync::Mutex<T> is invariant over the parameter T
    = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance

error: could not compile `cape_wallet` due to 2 previous errors
```

File: faucet/src/faucet.rs [hunk 1]
```diff
@@ -26,7 +26,10 @@ use jf_cap::{
     structs::{AssetCode, FreezeFlag},
 };
 use rand::distributions::{Alphanumeric, DistString};
-use reef::traits::Validator;
+
+// XXX looks wrong, but compilation fails if using reef::traits::Validator instead.
+use seahorse::reef::traits::Validator;
```
Comment (sveitser) [inline faucet/src/faucet.rs:31]: @jbearer Not sure if this makes sense. But it wouldn't compile with `reef::traits::Validator`.

Comment (sveitser) [reply faucet/src/faucet.rs:31]: Oh. Faucet seems to be using an outdated reef version.

File: faucet/src/faucet.rs [hunk 2]
```diff
@@ -26,11 +26,15 @@ use jf_cap::{
     structs::{AssetCode, FreezeFlag},
 };
 use rand::distributions::{Alphanumeric, DistString};
-use reef::traits::Validator;
+
+// XXX looks wrong, but compilation fails if using reef::traits::Validator instead.
```
Comment (jbearer) [inline faucet/src/faucet.rs]: Probably need to update `reef` to 0.2.2 in Cargo.toml?

File: faucet/src/faucet.rs [hunk 3]
```diff
@@ -275,12 +279,12 @@ async fn request_fee_assets(
...
 async fn spendable_records(
     wallet: &CapeWallet<'static, CapeBackend<'static, LoaderMetadata>>,
-    grant_size: u64,
+    grant_size: RecordAmount,
 ) -> impl Iterator<Item = RecordInfo> {
     let now = wallet.lock().await.state().txn_state.validator.now();
     wallet.records().await.filter(move |record| {
         record.ro.asset_def.code == AssetCode::native()
-            && record.ro.amount >= grant_size
+            && record.ro.amount >= grant_size.into()
```
Comment (jbearer) [inline faucet/src/faucet.rs:284]: You can also do `record.amount()` which returns a `RecordAmount`, then you wouldn't need `.into()`

**Comment by sveitser** [comment]
If anyone reviews this and feels like making modifications feel free to push directly to the branch. I have no more pending changes and won't be working on it again until tomorrow.

File: faucet/src/faucet.rs [hunk 4]
```diff
@@ -312,7 +316,7 @@ async fn break_up_records(state: FaucetState, mut wakeup: mpsc::Receiver<()>) {
                 );
             } else if !records
                 .into_iter()
-                .any(|record| record.ro.amount > 2 * state.grant_size)
+                .any(|record| RecordAmount::from(record.ro.amount) > state.grant_size * 2u64)
```
Comment (jbearer) [inline faucet/src/faucet.rs:316]: Same, `record.amount()` can be used instead of the conversion

File: wallet/src/backend.rs [hunk 5]
```diff
@@ -556,7 +556,10 @@ impl<'a, Meta: Serialize + DeserializeOwned + Send> CapeWalletBackend<'a>
         // Before the contract can transfer from our account, in accordance with the ERC20 protocol,
         // we have to approve the transfer.
         ERC20::new(erc20_code.clone(), eth.client())
-            .approve(eth.contract.address(), ro.amount.into())
+            .approve(
+                eth.contract.address(),
+                ro.amount.generic_into::<u128>().into(),
```
Comment (jbearer) [inline wallet/src/backend.rs:561]: Does `ro.amount.generic_into::<U256>()` not work?

File: wallet/src/ui.rs [hunk 6]
```diff
@@ -47,7 +48,7 @@ pub struct AssetDefinition {
     pub address_viewable: bool,
     pub amount_viewable: bool,
     pub blind_viewable: bool,
-    pub viewing_threshold: u64,
+    pub viewing_threshold: u128,
```
Comment (jbearer) [inline wallet/src/ui.rs:51]: Why `u128` instead of `RecordAmount`?

File: wallet/src/ui.rs [hunk 7]
```diff
@@ -418,7 +419,7 @@ pub struct WalletSummary {
 pub struct Record {
     pub address: UserAddress,
     pub asset: AssetCode,
-    pub amount: u64,
+    pub amount: Amount,
```
Comment (jbearer) [inline wallet/src/ui.rs:422]: Can we use `RecordAmount` instead of `Amount` for wallet stuff?

File: wallet/src/ui.rs [hunk 8]
```diff
@@ -806,7 +807,7 @@ pub mod sol {
         pub cred_pk: EdOnBN254Point,
         pub freezer_pk: EdOnBN254Point,
         pub reveal_map: U256,
-        pub reveal_threshold: u64,
+        pub reveal_threshold: u128,
```
Comment (jbearer) [inline wallet/src/ui.rs:810]: `RecordAmount`

File: wallet/src/ui.rs [hunk 9]
```diff
@@ -848,7 +849,7 @@ pub mod sol {
     #[ser_test(ark(false))]
     #[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
     pub struct RecordOpening {
-        pub amount: u64,
+        pub amount: u128,
```
Comment (jbearer) [inline wallet/src/ui.rs:852]: `RecordAmount`

File: wallet/src/wallet.rs [hunk 10]
```diff
@@ -134,7 +134,7 @@ pub trait CapeWalletExt<'a, Backend: CapeWalletBackend<'a> + Sync + 'a> {
         // it.
         cap_asset: AssetDefinition,
         dst_addr: UserAddress,
-        amount: u64,
+        amount: Amount,
```
Comment (jbearer) [inline wallet/src/wallet.rs:137]: To be consistent with the Seahorse API,
* use `RecordAmount`
* Allow conversions, e.g. `amount: impl Into<RecordAmount>`

Comment (jbearer) [reply wallet/src/wallet.rs:137]: Same goes for all the `Amount` parameters in this file

**Review by jbearer — CHANGES_REQUESTED** [review]
A few minor comments about conversions and consistency, but overall looks straightforward enough. Just remember to change the branch for Seahorse back to `release-0.2`

File: wallet/src/backend.rs (hunk 5 above)
Comment (sveitser) [reply wallet/src/backend.rs:561]: It will work with https://github.com/EspressoSystems/cap/pull/50

**Pull request merged**